test/lit/passes/type-merging.wast

```diff
@@ -138,25 +138,23 @@
   ;; CHECK:      (type $A (struct (field (ref null $A))))
   (type $A (struct_subtype (field (ref null $A)) data))
   (type $B (struct_subtype (field (ref null $A)) $A))
-  ;; CHECK:      (type $none_=>_none (func))
-
-  ;; CHECK:      (type $C (struct_subtype (field (ref null $A)) $A))
   (type $C (struct_subtype (field (ref null $B)) $A))
 
+  ;; CHECK:      (type $none_=>_none (func))
+
   ;; CHECK:      (func $foo (type $none_=>_none)
   ;; CHECK-NEXT:  (local $a (ref null $A))
   ;; CHECK-NEXT:  (local $b (ref null $A))
-  ;; CHECK-NEXT:  (local $c (ref null $C))
+  ;; CHECK-NEXT:  (local $c (ref null $A))
   ;; CHECK-NEXT:  (nop)
   ;; CHECK-NEXT: )
   (func $foo
     ;; $A will remain the same.
     (local $a (ref null $A))
     ;; $B can be merged into $A.
     (local $b (ref null $B))
-    ;; $C refines the field, so it cannot be merged. However, separately, in
-    ;; the type definition of $C, its field of type $B should become $A. That
-    ;; is, $B should no longer be used anywhere.
+    ;; $C refines the field, so normally it wouldn't be able to be merged, but
+    ;; since that refinement does not survive merging, we can still merge $C.
```

Comment:
1. Please modify this test to test what it used to (that a field is properly updated in a type that is not merged), or add a new one.
2. I am not sure I understand why the new result is valid. Is it only because `$C` is not used anywhere? (if so then I guess 1 could be achieved by using it)

Reply:
> I am not sure I understand why the new result is valid. Is it only because `$C` is not used anywhere? (if so then I guess 1 could be achieved by using it)

Note that this is the same result you would get if you had run the previous implementation of type merging for another iteration. Using `$C` in a cast would prevent the merge, yes. I'll update this test to use `$C` in a cast and test all the fun new merging capabilities separately.

Reply:
Actually, updating `$C` to use a cast would be testing something materially different. It would be testing that casts inhibit merging rather than that refining inhibits merging. I'll update this test another way.